core/arch/atomic_exports.h

```diff
@@ -625,34 +636,67 @@ atomic_dec_becomes_zero(volatile int *var)
+                                 "1: ldrex" suffix " r2, %0         \n\t" \
+                                 "   add" suffix " r2, r2, #1       \n\t" \
+                                 "   strex" suffix " r3, r2, %0     \n\t" \
+                                 "   cmp   r3, #0                   \n\t" \
+                                 "   bne   1b                       \n\t" \
+                                 : "=Q"(var) /* no offset for ARM mode */ \
+                                 :                                        \
                                  : "cc", "memory", "r2", "r3");
 #        define ATOMIC_INC_int(var) ATOMIC_INC_suffix("", var)
 #        define ATOMIC_INC_int64(var) ATOMIC_INC_suffix("d", var)
 #        define ATOMIC_INC(type, var) ATOMIC_INC_##type(var)
-#        define ATOMIC_DEC_suffix(suffix, var)                                     \
-            __asm__ __volatile__("   dmb ish                        \n\t"          \
-                                 "1: ldrex" suffix " r2, %0         \n\t"          \
-                                 "   sub" suffix " r2, r2, #1     \n\t"            \
-                                 "   strex" suffix " r3, r2, %0     \n\t"          \
-                                 "   cmp   r3, #0                   \n\t"          \
-                                 "   bne   1b                       \n\t"          \
-                                 "   cmp   r2, #0" /* for possible SET_FLAG use */ \
-                                 : "=Q"(var)       /* no offset for ARM mode */    \
+/* "res", the output, is assumed to be a single-byte bool. */
+#        define ATOMIC_INC_int_res(var, test, res)                                 \
+            __asm__ __volatile__("   dmb ish                \n\t"                  \
+                                 "1: ldrex r2, %0           \n\t"                  \
+                                 "   add r2, r2, #1         \n\t"                  \
+                                 "   strex r3, r2, %0       \n\t"                  \
+                                 "   cmp   r3, #0           \n\t"                  \
+                                 "   bne   1b               \n\t"                  \
+                                 "   cmp   r2, #0           \n\t"                  \
+                                 "   mov   r2, #1           \n\t"                  \
+                                 "   b" #test " 2f          \n\t"                  \
+                                 "   mov r2, #0             \n\t"                  \
+                                 "2: strb  r2, %1"                                 \
+                                 : "=Q"(var) /* no offs for ARM mode */, "=m"(res) \
                                  :                                                 \
                                  : "cc", "memory", "r2", "r3");
+#        define ATOMIC_INC_int_res_eq(var, res) ATOMIC_INC_int_res(var, eq, res)
```

Comment:
So the res is set to 1 if var was zero before the operation, and 0 otherwise?

The macro should describe what res is; or better s/res/was_var_zero/. The res in ATOMIC_INC_int_res can remain res I guess.